src/algo.rs

```diff
@@ -58,38 +56,47 @@ impl Incorporate for Vec<Profile> {
     fn incorporate(&mut self, candidate: Profile) -> bool {
         // The profiles are ordered in decreasing dep_time
         // The pivot is the element just before the candidate
-        let pivot = self.iter().rposition(|p| p.dep_time >= candidate.dep_time);
-        let mut earlier_profiles: Vec<Profile> = match pivot {
-            None => Vec::new(),
-            Some(position) => self.drain(position + 1..)
-                .filter(|p| candidate.dominates(p))
-                .collect(),
-        };
-
-        let incorporated = candidate.is_non_dominated(self.last());
-        if incorporated {
-            self.push(candidate);
+        match self.iter().rposition(|p| p.dep_time >= candidate.dep_time) {
+            Some(pivot) => {
+                if candidate.is_non_dominated(&self[pivot]) {
+                    let mut i = pivot + 1; // until we get drain_filter in stable
+                    while i < self.len() {
+                        if candidate.dominates(&self[i]) {
+                            self.swap_remove(i);
+                        } else {
+                            i += 1;
+                        }
+                    }
+                    self.push(candidate);
+                    self[pivot + 1..].sort_by(|a, b| b.dep_time.cmp(&a.dep_time));
+                    true
+                } else {
+                    false
+                }
+            }
+            None => {
+                self.push(candidate);
+                true
+            }
```

Comment:
I suspect you found a way to avoid the allocation.

That seems messy. You should abstract this dominance thing somewhere and using meaningful methods here.

Reply:
you are right, I believe it is now easier to understand.

Yet I’m not happy how to return the `bool` telling if the candidate was actually incorporated